### ISSUE TITLE
训练了10轮候，F1为0

### ISSUE THREAD

**Issue by ff202118** [SunnyHaze/IML-ViT#48]
![Image](https://github.com/user-attachments/assets/20790a0e-78bb-47e4-b926-69f672390e99)

由于我是一个新手，所以我没有太多的经验。我只有一张卡，所以没有使用分布式训练，同时我使用的是自己数据集。我有一个预感是因为自己数据集中 图像尺寸不固定，而且大小都在250px*250px左右，这是否对训练会有影响。

下面是我的参数列表。

`base_dir="./output_dir"
mkdir -p ${base_dir}

CUDA_VISIBLE_DEVICES=0 \
torchrun  \
    --standalone    \
    --nnodes=1     \
    --nproc_per_node=1 \
./train.py \
    --model IMLViTModel \
    --vit_pretrain_path ./pretrained-weights/mae_pretrain_vit_base.pth \
    --world_size 1 \
    --batch_size 2 \
    --data_path ./train1.json \
    --epochs 10 \
    --lr 1e-4 \
    --image_size 1024 \
    --if_padding \
    --min_lr 5e-7 \
    --weight_decay 0.05 \
    --edge_mask_width 7 \
    --test_data_path "./test_datasets.json" \
    --warmup_epochs 2 \
    --output_dir ${base_dir}/ \
    --log_dir ${base_dir}/ \
    --accum_iter 8 \
    --seed 42 \
    --test_period 4 \
2> ${base_dir}/error.log 1>${base_dir}/logs.log`

**Comment by SunnyHaze** [comment]
你好，这可能有多种因素：
1. 你的数据集本身尺寸较小，以padding策略引入了太多的黑像素，导致模型倾向于预测全黑。
2. 也有可能是数据集本身黑色区域较多。或者mask有问题。这里可以用`tensroboard --logdir ./`指令观看输出日志，尤其是mask相关的图片是否正常显示mask

改进建议：
1. 可以提高IML-ViT的`labmda`的权重，让篡改区域权重更大。具体来说可以在shell中加入如下代码`--edge_lambda 40`代表edge权重变大。
```shell
`base_dir="./output_dir"
mkdir -p ${base_dir}

CUDA_VISIBLE_DEVICES=0
torchrun
--standalone
--nnodes=1
--nproc_per_node=1
./train.py
--model IMLViTModel
--edge_lambda 40
--vit_pretrain_path ./pretrained-weights/mae_pretrain_vit_base.pth
--world_size 1
--batch_size 2
--data_path ./train1.json
--epochs 10
--lr 1e-4
--image_size 1024
--if_padding
--min_lr 5e-7
--weight_decay 0.05
--edge_mask_width 7
--test_data_path "./test_datasets.json"
--warmup_epochs 2
--output_dir ${base_dir}/
--log_dir ${base_dir}/
--accum_iter 8
--seed 42
--test_period 4
2> ${base_dir}/error.log 1>${base_dir}/logs.log`
```
2. 可以考虑扩大数据集的尺寸，或者直接引入一些CASIAv2尺寸较大的图片辅助收敛。
3. 作为ViT模型，10个Epoch是比较难收敛的，总体应该还是要多一些，不过是建立在有收敛倾向的前提下。

**Comment by ff202118** [comment]
好的，我尝试一下，谢谢你



---原始邮件---
发件人: "Ma, ***@***.***&gt;
发送时间: 2025年4月26日(周六) 凌晨1:17
收件人: ***@***.***&gt;;
抄送: "Jiaqi ***@***.******@***.***&gt;;
主题: Re: [SunnyHaze/IML-ViT] 训练了10轮候，F1为0 (Issue #48)


 SunnyHaze left a comment (SunnyHaze/IML-ViT#48)
 
你好，这可能有多种因素：
  
你的数据集本身尺寸较小，以padding策略引入了太多的黑像素，导致模型倾向于预测全黑。
 
也有可能是数据集本身黑色区域较多。或者mask有问题。这里可以用tensroboard --logdir ./指令观看输出日志，尤其是mask相关的图片是否正常显示mask
  
改进建议：
  
可以提高IML-ViT的labmda的权重，让篡改区域权重更大。具体来说可以在shell中加入如下代码--edge_lambda 40代表edge权重变大。
  `base_dir="./output_dir" mkdir -p ${base_dir}  CUDA_VISIBLE_DEVICES=0 torchrun --standalone --nnodes=1 --nproc_per_node=1 ./train.py --model IMLViTModel --edge_lambda 40 --vit_pretrain_path ./pretrained-weights/mae_pretrain_vit_base.pth --world_size 1 --batch_size 2 --data_path ./train1.json --epochs 10 --lr 1e-4 --image_size 1024 --if_padding --min_lr 5e-7 --weight_decay 0.05 --edge_mask_width 7 --test_data_path "./test_datasets.json" --warmup_epochs 2 --output_dir ${base_dir}/ --log_dir ${base_dir}/ --accum_iter 8 --seed 42 --test_period 4 2&gt; ${base_dir}/error.log 1&gt;${base_dir}/logs.log`
  
可以考虑扩大数据集的尺寸，或者直接引入一些CASIAv2尺寸较大的图片辅助收敛。
 
作为ViT模型，10个Epoch是比较难收敛的，总体应该还是要多一些，不过是建立在有收敛倾向的前提下。

**Comment by SunnyHaze** [comment]
实在不行可以尝试用resize 512的方式训练IML-ViT，不过可能要略微修改一下结构适应512的输入。

**Comment by ff202118** [comment]
![Image](https://github.com/user-attachments/assets/c8bbb62d-59db-4491-86af-ca407072d0e1)
首先感谢你的耐心指导。我按照你的说法，进行了尝试，但是我发现测试的F1分数居然大于了1，请问你是否使用了百分制来表达？（我看了源代码似乎并没有，那为什么会出现大于1呢）。
我设置的是参数是：
base_dir="./output_dir"
mkdir -p ${base_dir}

CUDA_VISIBLE_DEVICES=0 \
torchrun  \
    --standalone    \
    --nnodes=1     \
    --nproc_per_node=1 \
./train.py \
    --model IMLViTModel \
    --vit_pretrain_path ./pretrained-weights/mae_pretrain_vit_base.pth \
    --world_size 1 \
    --batch_size 2 \
    --data_path train.json \
    --epochs 50 \
    --lr 1e-4 \
    --image_size 512 \
    --if_padding \
    --min_lr 5e-7 \
    --weight_decay 0.05 \
    --edge_mask_width 7 \
    --test_data_path "test_datasets.json" \
    --warmup_epochs 2 \
    --output_dir ${base_dir}/ \
    --log_dir ${base_dir}/ \
    --accum_iter 8 \
    --seed 42 \
    --test_period 4 \
    --edge_lambda 40\
2> ${base_dir}/error.log 1>${base_dir}/logs.log

**Comment by SunnyHaze** [comment]
你好，这个问题之前也有人注意到了，请务必以括号内的数值，以及每个epoch最后测试归并完成的数值为准。

具体解答参考此issue：
https://github.com/scu-zjz/IMDLBenCo/issues/90

**Comment by SunnyHaze** [comment]
我们在最新的BenCo版本修正了这个问题，感谢你提出的建议。请更新benco版本到最新的`v0.1.32`以部署相应的更新。

具体内容参考此release：
https://github.com/scu-zjz/IMDLBenCo/releases/tag/v0.1.32

**Comment by ff202118** [comment]
感谢你的指导！

**Comment by SunnyHaze** [comment]
不客气！祝科研顺利！

一定要用v0.1.33以及以后的版本哈，v0.1.32有问题。参考如下release
https://github.com/scu-zjz/IMDLBenCo/releases/tag/v0.1.33

可以的话一定要cite我们的两篇工作，感谢:)
